### ISSUE TITLE
Add JTAG support

### ISSUE THREAD

**Issue by P33M** [raspberrypi/picoprobe#44]
Given that we already swap probe programs in and out of PIO on the fly, it would be easy enough to add JTAG support.

How does PIO assist in this scenario - is the DAP protocol sufficiently high-level that we could accelerate scan-IR and scan-DR ops?

**Comment by diggit** [comment]
DAP JTAG API is pretty low level.
https://arm-software.github.io/CMSIS_5/DAP/html/group__DAP__JTAG__Sequence.html
PIO could handle this configurable 1~64 bit "SPI".

**Comment by Patronics** [comment]
Worth noting that there seem to be a few other projects that implement JTAG on the Pi Pico, perhaps some of that functionality could be ported over relatively easily?

There's https://github.com/phdussud/pico-dirtyJtag/

https://github.com/kholia/xvc-pico/

and https://git.lain.faith/sys64738/DragonProbe.git

**Comment by Terstegge** [comment]
You could try out my version of a firmware for the RPi Debug Probe board:
https://github.com/Terstegge/picoprobepp
I tested the firmware with a MSP432 Launchpad using a JTAG connection - at least I could write the flash and do some debugging.

**Comment by Himanshu-4** [comment]
i had developed the JTAG support in existing pi pico firmware, i have tested on esp32 and working perfectly. 
the pio version of JTAG is still under development.
You can check it out on https://www.linkedin.com/feed/update/urn:li:activity:7436369988216569856/

**Comment by Patronics** [comment]
In the interest of ease of access to information, the above LinkedIn post links to the following fork of the project: https://github.com/Himanshu-4/debugprobe.

I haven't investigated this fork in much detail, though it is interesting to note the sequence of large commits in rapid succession (all commits within an 8 minute span!), some of which are adding unnecessary logging files, before then removing them, some small unnecessary changes such as which pins are tied to the debug pins, and the entirely-commented out src/jtag.c file. Definitely seems to be some substantial AI involvement in producing these changes, but it's not my place to assign a value judgement to that, or to decide whether that complies with the policies of this project.

**Comment by P33M** [comment]
> Definitely seems to be some substantial AI involvement in producing these changes,

Agreed. Notwithstanding the other issues with inclusion of extraneous cruft, without clear attribution, it's not suitable for merging into an open-source project.

**Comment by Himanshu-4** [comment]
Hi @Patronics, the PIO version of the JTAG is still under development, i will try to implement it ASAP and according to the repo policies. then will raise a pull request.
